src/HW_models/NRF_EGU.c

```diff
@@ -4,10 +4,138 @@
+// }
+FOR_EACH(NRF_EGU_0_TASK_TRIGGER_DEFINE, ( ), CHANNEL_LIST)
+FOR_EACH(NRF_EGU_1_TASK_TRIGGER_DEFINE, ( ), CHANNEL_LIST)
+FOR_EACH(NRF_EGU_2_TASK_TRIGGER_DEFINE, ( ), CHANNEL_LIST)
+FOR_EACH(NRF_EGU_3_TASK_TRIGGER_DEFINE, ( ), CHANNEL_LIST)
+FOR_EACH(NRF_EGU_4_TASK_TRIGGER_DEFINE, ( ), CHANNEL_LIST)
+FOR_EACH(NRF_EGU_5_TASK_TRIGGER_DEFINE, ( ), CHANNEL_LIST)
+
+#define EGU0_EVENTS_TRIGGERED(N, _) EGU0_EVENTS_TRIGGERED_ ## N
+#define EGU1_EVENTS_TRIGGERED(N, _) EGU1_EVENTS_TRIGGERED_ ## N
+#define EGU2_EVENTS_TRIGGERED(N, _) EGU2_EVENTS_TRIGGERED_ ## N
+#define EGU3_EVENTS_TRIGGERED(N, _) EGU3_EVENTS_TRIGGERED_ ## N
+#define EGU4_EVENTS_TRIGGERED(N, _) EGU4_EVENTS_TRIGGERED_ ## N
+#define EGU5_EVENTS_TRIGGERED(N, _) EGU5_EVENTS_TRIGGERED_ ## N
+
+const ppi_event_types_t PPI_EGU_EVENTS[EGU_COUNT][EGU0_CH_NUM] = {
+	{
+		//EGU0_EVENTS_TRIGGERED_0, ...
+		//EGU0_EVENTS_TRIGGERED_15
+		LISTIFY(EGU0_CH_NUM, EGU0_EVENTS_TRIGGERED, (,))
+	},
+	{
+		LISTIFY(EGU0_CH_NUM, EGU1_EVENTS_TRIGGERED, (,))
+	},
+	{
+		LISTIFY(EGU0_CH_NUM, EGU2_EVENTS_TRIGGERED, (,))
+	},
+	{
+		LISTIFY(EGU0_CH_NUM, EGU3_EVENTS_TRIGGERED, (,))
+	},
+	{
+		LISTIFY(EGU0_CH_NUM, EGU4_EVENTS_TRIGGERED, (,))
+	},
+	{
+		LISTIFY(EGU0_CH_NUM, EGU5_EVENTS_TRIGGERED, (,))
+	},
+};
+
+/**
+ * @brief Get Interrupt Id of corresponding EGU peripheral
+ *
+ * @param p_reg NRF_EGU_Type pointer
+ * @return IRQn for p_reg
  */
+#define REG_TO_IRQ(p_reg)   (REG_TO_ID(p_reg) + SWI0_EGU0_IRQn)
+
+/**
+ * @brief Check if channel has enabled interrupts
+ *
+ * @param p_reg NRF_EGU_Type pointer
+ * @param channel uint8_t channel number to check
+ * @return True if interrupt is enabled for channel in p_reg, otherwise return False
+ */
+#define CHANNEL_INTERRUPT_FLAG(p_reg, channel)    (p_reg->INTEN & nrf_egu_channel_int_get(channel))
+
+
+NRF_EGU_Type NRF_EGU_regs[EGU_COUNT];
+// prevent enabling already enabled IRQ
+bool EGU_interrupt_enabled[EGU_COUNT];
+
+void nrf_egu_regw_sideeffects_TASKS_TRIGGER(NRF_EGU_Type *p_reg)
+{
+	// move tasks to triggers
+	memcpy(p_reg->EVENTS_TRIGGERED, p_reg->TASKS_TRIGGER, sizeof(p_reg->EVENTS_TRIGGERED));
+	// task has been moved to trigger, zero out the register for future use
+	memset(p_reg->TASKS_TRIGGER, 0, sizeof(p_reg->TASKS_TRIGGER));
+	nrf_egu_regw_sideeffects_EVENTS_TRIGGERED_INTEN(p_reg);
+}
+
+void nrf_egu_regw_sideeffects_EVENTS_TRIGGERED_INTEN(NRF_EGU_Type *p_reg)
+{
+	for (uint8_t i = 0; i < nrf_egu_channel_count(p_reg); i++) {
+		if (CHANNEL_INTERRUPT_FLAG(p_reg, i) && p_reg->EVENTS_TRIGGERED[i]) {
+			// if there is some task, and interrupt is enabled - set irq
+			if (!EGU_interrupt_enabled[REG_TO_ID(p_reg)]) {
+				EGU_interrupt_enabled[REG_TO_ID(p_reg)] = true;
+				hw_irq_ctrl_set_irq(REG_TO_IRQ(p_reg));
+			}
+			return;
+		}
+	}
+	// if there are not active tasks with enabled interrupt - clear irq
+	if (EGU_interrupt_enabled[REG_TO_ID(p_reg)]) {
+		EGU_interrupt_enabled[REG_TO_ID(p_reg)] = false;
+		hw_irq_ctrl_clear_irq(REG_TO_IRQ(p_reg));
```

Comment:
This looks weird. Interrupts once pended to the IRQ controller are normally only cleared by the interrupt handler (but this is your HW :) )

Reply:
What do You mean ? When I tested the real hardware, if I did not call nrf_egu_event_clear in the irq handler, the irq would be triggered forever, so the act of writing 0 to Event register, has sideeffect of disabling IRQ (if there is no other pending event with interrupts enabled)

Reply:
We are talking a bit pass each other :). One thing is the interrupt line from the peripheral to the interrupt controller (the NVIC) another the internal status of the interrupt controller:
An NVIC once it sees an interrupt raise "pends" the interrupt internally. And it will clear that pending status once it causes the CPU core to vector into the corresponding handler. 
(Now, if the interrupt line from the peripheral is still high at the end of the interrupt handler execution, the interrupt will be pended right away again inside the NVIC and the NVIC will cause the interrupt handler to be executed again but that is another matter)
What you are clearing here is the equivalent of the NVIC internal status.

------------

Independently of this, I did not model in general interrupts from peripherals to the NVIC being level interrupts (so there is no good way of representing that an interrupt is left up after the interrupt handler returns). You can imagine that in this model the peripherals are only sending a pulse to the NVIC when an interrupt is triggered. 
Modelling level interrupts would require some changes in the interrupt controller<->peripheral API. But I hope it does not matter a lot as any proper interrupt handler will clear the irq status (EVENT) in the peripheral so it does not retrigger. Meaning I hope you don't need to model the situation where an interrupt handler forgets to clear the corresponding EVENT.

Reply:
Level interrupts have been added here:
https://github.com/BabbleSim/ext_NRF52_hw_models/commit/65bc5305d432c08e24a3f343006d1e7deaff4908
Theis API in the IRQ controller is:
`hw_irq_ctrl_raise_level_irq_line()`
`hw_irq_ctrl_lower_level_irq_line()`
(Note that it only supports "normal use" cases, I did not add support to SW trying to unpend an interrupt thru the NVIC registers while the line is still high)

It needs this PR to the interrupt handler in the nr52_bsim to get it into use:
https://github.com/zephyrproject-rtos/zephyr/pull/49420